timemachine/fe/interpolate.py

```diff
@@ -1,217 +1,120 @@
-
-    src_kv = _add_dst_to_src_torsion(src_idxs, src_params, dst_idxs, dst_params)
-    dst_kv = _add_dst_to_src_torsion(dst_idxs, dst_params, src_idxs, src_params)
-
-    assert src_kv.keys() == dst_kv.keys()
-
-    res = set()
-
-    for key in src_kv.keys():
-        idxs, period = key
-        assert idxs[0] < idxs[-1]
-        src_k, src_phase = src_kv[key]
-        dst_k, dst_phase = dst_kv[key]
-        res.add((idxs, (src_k, src_phase, period), (dst_k, dst_phase, period)))
-
-    return res
-
-
-def _add_dst_to_src_nonbonded(src_idxs, src_params, dst_idxs):
-    assert len(set((src_idxs))) == len(src_idxs)
-    assert len(set((dst_idxs))) == len(dst_idxs)
-
-    src_kv = dict(zip(src_idxs, src_params))
-    for idxs in dst_idxs:
-        assert idxs[0] < idxs[-1]
-        if idxs not in src_kv:
-            src_kv[idxs] = (0, 0, 0)
-
-    return src_kv
-
-
-def align_nonbonded_idxs_and_params(src_idxs, src_params, dst_idxs, dst_params):
-
-    # sanitize
-    src_idxs = [tuple(p) for p in src_idxs]
-    src_params = [tuple(p) for p in src_params]
-    dst_idxs = [tuple(p) for p in dst_idxs]
-    dst_params = [tuple(p) for p in dst_params]
-
-    src_kv = _add_dst_to_src_nonbonded(src_idxs, src_params, dst_idxs)
-    dst_kv = _add_dst_to_src_nonbonded(dst_idxs, dst_params, src_idxs)
-    assert src_kv.keys() == dst_kv.keys()
-
-    res = set()
-
-    for k in src_kv.keys():
-        res.add((tuple(k), src_kv[k], dst_kv[k]))
-
-    return res
-
-
-# special case for torsions
-def _add_dst_to_src_chiral_atom(src_idxs, src_params, dst_idxs):
-    src_kv = dict()
-
-    for idxs, k in zip(src_idxs, src_params):
-        src_kv[idxs] = k
-
-    for idxs in dst_idxs:
-        if idxs not in src_kv:
-            src_kv[idxs] = 0
-
-    return src_kv
-
-
-def align_chiral_atom_idxs_and_params(src_idxs, src_params, dst_idxs, dst_params):
-
-    # sanitize
-    src_idxs = [tuple(p) for p in src_idxs]
-    dst_idxs = [tuple(p) for p in dst_idxs]
-
-    src_kv = _add_dst_to_src_chiral_atom(src_idxs, src_params, dst_idxs)
-    dst_kv = _add_dst_to_src_chiral_atom(dst_idxs, dst_params, src_idxs)
-    assert src_kv.keys() == dst_kv.keys()
-
-    res = set()
-
-    for k in src_kv.keys():
-        res.add((tuple(k), src_kv[k], dst_kv[k]))
-
-    return res
-
-
-def _add_dst_to_src_chiral_bond(src_idxs, src_params, src_signs, dst_idxs, dst_params, dst_signs):
-    src_kv = dict()
-
-    for idxs, k, sign in zip(src_idxs, src_params, src_signs):
-        assert idxs[0] < idxs[-1]
-        key = (idxs, sign)
-        src_kv[key] = k
-
-    for idxs, k, sign in zip(dst_idxs, dst_params, dst_signs):
-        assert idxs[0] < idxs[-1]
-        key = (idxs, sign)
-        if key not in src_kv:
-            src_kv[key] = 0
-
-    return src_kv
+align_nonbonded_idxs_and_params = partial(align_idxs_and_params, make_default=lambda _: (0, 0, 0))
+align_chiral_atom_idxs_and_params = partial(align_idxs_and_params, make_default=lambda _: 0)
```

Comment:
we may need to do one more check on the canonicalization of chiral terms (missing in master as well) for alignment purposes:

```
(0,1,2,3)
(0,3,2,1)
(0,2,1,3)
(2,1,3,0)
(3,1,0,2)
(1,3,2,0)
(3,0,2,1)
(1,2,0,3)
(2,0,1,3)
```

I'm not sure if we actually do this correctly!

Reply:
nevermind - this should be taken care of by the canonicalization routine - will double check next week

Reply:
[relevant bits](https://github.com/proteneer/timemachine/blob/master/timemachine/fe/single_topology_v3.py#L355)

Okay so the code currently doesn't quite do full canonicalization, but it's probably okay (and intended behavior). We define the chiral restraints as being "centered" at the `0th` index atom:

```
(0,1,2,3)
(0,3,2,1)
(0,2,1,3)
```
is distinct from 
```
(1,0,2,3)
(1,3,2,0)
(1,2,0,3)
```
(even though they result in the same sign). Closing this and sorry for the confusion.